### DISCUSSION TITLE
groupBy and Method Illuminate\Database\Eloquent\Collection::getKey does not exist

### DISCUSSION THREAD

**Discussion by KeironLowe** [livewire/livewire#3066] [7. Bugs (archived)]
### Description
When binding the value of an Eloquent groupBy method into a Livewire component, you receive the following error. `Method Illuminate\Database\Eloquent\Collection::getKey does not exist`.

The same code works fine when used in a non-livewire component, such as a component or view composer.

### Steps to reproduce

1. Create a livewire component
2. In the `mount` method, try bind a property to `User::all()->groupBy('created_at');`

**Context**
- Livewire version: [e.g. 11.0]
- Laravel version: [e.g. 7.0.0]

**Comment by stevebauman** [comment]
Hi @KeironLowe,

I believe this is due to the `created_at` being a Carbon instance, and not a string or integer.

You must do this instead:

```php
public function mount()
{
    $this->users = User::all()->groupBy(function ($user) {
        // Group the users by creation day.
        return $user->created_at->toDateString();
    })->all();
}
```

Using your method wouldn't group a user properly anyway, since it would effectively group them by every second, since Carbon instances get transformed into DateTime strings when casted to strings.

Hope this helps! 👍

**Comment by stevebauman** [comment]
I have confirmed that this is indeed not a bug with Livewire -- it is Laravel related.

You can test this simply by running this in your `routes/web.php`:

```php
// routes/web.php

dd(\App\User::all()->groupBy('created_at'));
```

Which will generate an exception, since the `created_at` attribute is a Carbon instance, which collections cannot cast.

**Comment by KeironLowe** [comment]
Thanks a lot, my example was a simplified example. In my code, I actually was grouping by a string and not a carbon instance. I've got it working and I could be wrong but I think the issue wasn't Carbon.

The code that throws the error is this...

```
$this->jobs = Job::forWeek()->get()->groupBy(function ($job) {
    return date("l", strtotime($job->start_date));
});
```

So it's fetching the jobs for the current week, and grouping them by day. This throws the mentioned error, but adding `->all()` to the end fixes it. With that, it's no longer a collection but an array of collections, while without it was returning the eloquent collection. But returning an Eloquent collection should work shouldn't it?

**Comment by stevebauman** [comment]
Ah yeah you’re right... I’m able to reproduce this as well. You could re-collect them via `collect()` in the render method or in the view, but it’s not ideal. I’m going to check into the source and see if this is patchable.

**Comment by ahawlitschek** [comment]
I am receiving the same error when trying to use group by with an integer field.
Really strange

**Comment by spiekerr** [comment]
Any updates?

**Comment by ssarljames** [comment]
I have the same error. I fix it by having an array value instead of collection after grouping.

```php
public function mount()
{
    $this->eventTypes = EventType::all()->groupBy(function ($eventType) {
        return $eventType->group;
    })->toArray();
} 
```

I think the issue here is the casting of data.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by rohithmeethal** [comment]
REOPEN

**Comment by rohithmeethal** [comment]
Issue still exists when we use collection that has groupBy applied.

**Comment by simple-hacker** [comment]
I can confirm this is still not working.  My temporary solution is to wrap it with `collect()` which is an instance of `Illuminate\Support\Collection` instead of `Illuminate\Database\Eloquent\Collection`

`collect(Permission::all()->groupBy('group'))`

**Comment by marky291** [comment]
Issue still exists.

workaround for me was using “where()” on the ungrouped collection. To get the needed grouping in view.

**Comment by austenc** [comment]
@rohithmeethal @simple-hacker @marky291 

Could one of you make a repository that illustrates this problem? If so, we'd be happy to look into it! Thank you!

**Comment by guoxiangke** [comment]
same error: @austenc 
![image](https://user-images.githubusercontent.com/1160703/111126574-cec84580-85ad-11eb-868e-f452529e7077.png)

**Reply by SamirMaharjan** [reply]
this worked for me Thanks

**Comment by joshhanley** [comment]
Livewire doesn't support using `groupBy` on eloquent collections.

The reason for that is, they way Livewire dehydrates and rehydrates eloquent collections, is by getting all of the keys of the items in the collection, and sending them to the front end (but not the rest of the data).

When it the eloquent collection is restored on the next request, those keys are used to get the records from the database, rehydrating the eloquent collection with all the records it had before.

But when you do a groupBy, Livewire can't serialise the eloquent collection, as it can't call `getKey` on each of the items as `groupBy` puts them into nested collections (hence the error that is being displayed).

Livewire has no way of knowing how the collection has been grouped.

To work around this, you could use computed properties, so lets say we have a list of users that we want grouped by name, first just get all the users and assign the standard eloquent collection to a public property 

```php
public $users;

public function mount() {
    $this->users = User::all();
}
```

Then you could have a computed property `$this->usersByName` and use this in your blade view
```php
public function getUsersByNameProperty {
    return $this->users->groupBy('name');
}
```

Hope this helps!

**Comment by marky291** [comment]
@joshhanley forgive me if I am wrong but that code snippet would be an eloquent collection and would cause the error in question.

**Comment by joshhanley** [comment]
@marky291 yep, but computed properties don't get serialised, so you won't run into that issue.

Instead the computer property will compute the group by when it gets used in your blade view. But all the data is still serialised using the non-grouped public users property (as it doesn't touch the users property, just uses it).

Computed properties also get memoized within a request too, so it will only compute it the first time it gets called (within a single request that is).

In your blade view, you would use the computed property, not the public property, so something like this

```html
@foreach ($this->usersByName as $name => $nameGroup)
    // Do stuff with group here
@endforeach
```

Hope that make sense 🙂

**Comment by pactode** [comment]
Another valid option is to use the `toBase` method. Example:
```php
$query->get()->groupBy('some_column')->toBase();
```

**Comment by jjdejong** [comment]
> Another valid option is to use the `toBase` method. Example:
> 
> ```
> $query->get()->groupBy('some_column')->toBase();
> ```

Worked for me. Appears to be the most elegant solution.

**Comment by kokoshneta** [comment]
This was not an easy explanation to find. I was passing a manually created nested collection and was getting this error.

Since there is already [a red-exclamation-mark notice](https://laravel-livewire.com/docs/2.x/properties#important-notes) in the documentation, it would be very useful to add an extra line to that, to the effect that properties of type `(Eloquent)Collection` **must** be non-nested and contain Eloquent models for the serialisation to work.

**Comment by ChrystlyWiafe** [comment]
There is really no problem with Laravel or Livewire; just follow this and you should be fine. 

The `->groupby('created_at')` return a collection `Illuminate\Database\Eloquent\Collection` .  so called the `all()` prop and this should work just fine.

That is to say that change this `User::all()->groupBy('created_at');` to this `User::all()->groupBy('created_at')->all();`

Ref: [https://laravel.com/docs/9.x/collections#method-groupby](url)
